src/annotator/util/buckets.js

```diff
@@ -0,0 +1,248 @@
+import { getBoundingClientRect } from '../highlighter';
+
+/**
+ * @typedef {import('../../types/annotator').Anchor} Anchor
+ */
+
+/**
+ * A structured Array representing either the top (`startOrEnd` = 1) or the
+ * bottom (`startOrEnd` = -1) of an anchor's highlight bounding box.
+ *
+ * @typedef {[pixelPosition: number, startOrEnd: (-1 | 1), anchor: Anchor]} PositionPoint
+ */
+
+/**
+ * An object containing information about anchor highlight positions
+ *
+ * @typedef PositionPoints
+ * @prop {Anchor[]} above - Anchors that are offscreen above
+ * @prop {Anchor[]} below - Anchors that are offscreen below
+ * @prop {PositionPoint[]} points - PositionPoints for on-screen anchor
+ *   highlights. Each highlight box has 2 PositionPoints (one for top edge
+ *   and one for bottom edge).
+ */
+
+/**
+ * @typedef BucketInfo
+ * @prop {Array<Anchor[]>} buckets
+ * @prop {number[]} index - Array of (pixel) positions of each bucket
+ */
+
+// FIXME: Temporary duplication of size constants between here and BucketBar
+const BUCKET_SIZE = 16; // Regular bucket size
+const BUCKET_NAV_SIZE = BUCKET_SIZE + 6; // Bucket plus arrow (up/down)
+const BUCKET_TOP_THRESHOLD = 115 + BUCKET_NAV_SIZE; // Toolbar
+
+// TODO!! This is an option in the plugin right now
+const BUCKET_GAP_SIZE = 60;
+
+/**
+ * Find the closest valid anchor in `anchors` that is offscreen in the direction
+ * indicated.
+ *
+ * @param {Anchor[]} anchors
+ * @param {'up'|'down'} direction
+ * @return {Anchor|null} - The closest anchor or `null` if no valid anchor found
+ */
+export function findClosestOffscreenAnchor(anchors, direction) {
```

Comment:
This function was already present on `master`, but this module has been moved and types added here.